### PR TITLE
Adopt the extended state in DataCoord

### PR DESCRIPTION
**Pull request by XuanYang-cn**
ChannelWatchInfo with ToWatch, ToRelease will trigger a timer.
ChannelManager now reacts to different ChannelWatch states.

- WatchSuccess > log this info

- WatchFailure/WatchTimeout > ToRelease

- ReleaseSuccess > Delete, reassign if not from DropCollection

- ReleaseFailure/ReleaseTimeout > Cleanup subscription and Delete,
  reassgin if not from DropCollection.

Some Notes:
1. Reassignment will add this channel to buffer if there's only one node.

See also: #15846

Signed-off-by: yangxuan <xuan.yang@zilliz.com>

### CODE REVIEW

File: internal/datacoord/channel_manager.go [hunk 1]
```diff
@@ -438,3 +521,284 @@ func (c *ChannelManager) findChannel(channelName string) (int64, *channel) {
...
+	}
+}
+
+// Release writes ToRlease channel watch states for a channel
+func (c *ChannelManager) Release(nodeID UniqueID, channelName string, callback func(string) error) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.dropCallback == nil && callback != nil {
+		c.dropCallback = callback
```
Comment (xiaofan-luan) [inline internal/datacoord/channel_manager.go:715]: what if server crash?
How do we know if c.dropCallback exists or not?

File: internal/datacoord/channel_manager.go [hunk 2]
```diff
@@ -98,14 +112,19 @@ func NewChannelManager(
...
+func (c *ChannelManager) Startup(ctx context.Context, nodes []int64) error {
 	channels := c.store.GetNodesChannels()
 	// Retrieve the current old nodes.
 	oNodes := make([]int64, 0, len(channels))
 	for _, c := range channels {
 		oNodes = append(oNodes, c.NodeID)
 	}
 
+	// Process watch states for old nodes.
+	if err := c.checkOldNodes(oNodes); err != nil {
```
Comment (xiaofan-luan) [inline internal/datacoord/channel_manager.go:124]: what happened if checkOldNodes is in ChannelWatchState_ToWatch but during the init datanode change it ChannelWatchState_WatchFailure or WatchSuccess before we watch?

Should we start stateChecker before checkOldNodes but add a lock to hold stateChecker?

File: internal/datacoord/channel_manager.go (hunk 1 above)
Comment (XuanYang-cn) [reply internal/datacoord/channel_manager.go:715]: This callback is set at s.DropVChannel when calling `channelManager.Release(xxx, xxx, callback)`, and only when channel's collections is dropped, it'll need a callback later.

File: internal/datacoord/cluster_test.go [hunk 3]
```diff
@@ -23,16 +23,35 @@ import (
 
 	"github.com/golang/protobuf/proto"
 	"github.com/milvus-io/milvus/internal/kv"
-	memkv "github.com/milvus-io/milvus/internal/kv/mem"
```
Comment (congqixia) [inline internal/datacoord/cluster_test.go:26]: It's recommended to use embeded etcd instead other KVs in ut.

**Pull request merged**